build-tools/src/testFixtures/java/org/elasticsearch/gradle/internal/test/InMemoryJavaCompiler.java

```diff
@@ -0,0 +1,173 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License
+ * 2.0 and the Server Side Public License, v 1; you may not use this file except
+ * in compliance with, at your election, the Elastic License 2.0 or the Server
+ * Side Public License, v 1.
+ */
+
+package org.elasticsearch.gradle.internal.test;
+
+import java.io.ByteArrayOutputStream;
+import java.io.IOException;
+import java.io.OutputStream;
+import java.net.URI;
+import java.util.List;
+import java.util.Map;
+import java.util.function.Supplier;
+import java.util.stream.Collectors;
+
+import javax.tools.FileObject;
+import javax.tools.ForwardingJavaFileManager;
+import javax.tools.JavaCompiler;
+import javax.tools.JavaCompiler.CompilationTask;
+import javax.tools.JavaFileManager;
+import javax.tools.JavaFileObject;
+import javax.tools.JavaFileObject.Kind;
+import javax.tools.SimpleJavaFileObject;
+import javax.tools.ToolProvider;
+
+/**
+ * An in-memory java source code compiler. InMemoryJavaCompiler can be used for compiling source
+ * code, represented as a CharSequence, to byte code, represented as a byte[].
+ *
+ * <p> The compiler will not use the file system at all, instead using a ByteArrayOutputStream for
+ * storing the byte code.
+ *
+ * <p> Example:
+ * <pre>{@code
+ *     Map<String, CharSequence> sources = Map.of(
+ *       "module-info",
+ *       """
+ *       module foo {
+ *         exports p;
+ *       }
+ *       """,
+ *       "p.Foo",
+ *       """
+ *       package p;
+ *       public class Foo implements java.util.function.Supplier<String> {
+ *        @Override public String get() {
+ *          return "Hello World!";
+ *         }
+ *       }
+ *       """
+ *     );
+ *     Map<String, byte[]> result = compile(sources);
+ * }</pre>
+ */
+public class InMemoryJavaCompiler {
```

Comment:
can we use anything from groovy to compile? maybe localclasses? bytebuddy?

Reply:
Not with the same convenience I guess as you can load groovy classes from a string, but you require writing it back to a jar as far as I understand?

Reply:
yes, I would have to get their byte code to write them to jar

Reply:
👍